crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -118,6 +128,11 @@ impl<DB: Database, C: Consensus, EF: ExecutorFactory> BlockchainTree<DB, C, EF>
                 last_canonical_hashes.last().cloned().unwrap_or_default()
             };
 
+        // size of the broadcast is double of max reorg depth because at max reorg depth we can have
+        // send at least N block at the time.
+        let (new_block_notication_sender, _) =
+            tokio::sync::broadcast::channel(2 * max_reorg_depth as usize);
```

Comment:
buffer capacity of broadcast is fixed and unlike regular mpsc does not grow with the number of senders.

I'm not certain this will always be enough under all circumstances. I'm not sure when a slot is cleared in a broadcast channel. when all receivers received that value? 
With a large number of subscriptions, this could perhaps lead to some drift and 
looking at the broadcast `send` function it looks like it overwrites the oldest value in the buffer.

imo it is safe to use an unbounded sender per subscription so all subscriptions will always see all blocks, that's why I'd prefer a list of senders over a broadcast channel.

Reply:
There is a lagging section that covers this: https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html#lagging
You have a limit of items in broadcast and if limit is hit oldest value is dropped. Items are shared between receivers.

The limit is quite big and in the usual eth network, it would take around ~2*13min to be filled (new blocks come at 13s period, usual reorgs are one block).

Maybe the question for this is what limits are on subscriber, and when does it get dropped. I assume it needs live TCP connection so at least there is a timeout on that. And from here: https://geth.ethereum.org/docs/interacting-with-geth/rpc/pubsub it seems that geth has a limit on the number of events

I feel that notification of new blocks is better suited for broadcasting than one to many subscribers, especially if we expect a lot of subscribers. But if you think it will be easier to integrate with mpsc we can switch it, it is a good enough reason, I don't have a strong stance against switching.

Reply:
thanks for the link!

I see, this sounds very reasonable to me now, 

> ~2*13min to be filled

this should be more than enough for all subscription task to keep up with new values

Reply:
If you find in future that mpsc is better suited, be free to change it